python/lib/core/dmod/core/dataset.py

```diff
@@ -168,21 +168,54 @@ def _serialize_datetime(self: "Dataset", value: Optional[datetime]) -> Optional[
 
     def __init__(self, manager: DatasetManager = None, **kwargs):
         super().__init__(**kwargs)
-
-        self._manager = manager if isinstance(manager, DatasetManager) else None
-
-        if manager is not None:
-            # pydantic will not validate this, so we need to check it
-            if not isinstance(manager.uuid, UUID):
-                raise ValueError(f"Expected UUID got {type(manager.uuid)}")
-            self.manager_uuid = manager.uuid
+        self.set_manager(manager)
+
+    def __eq__(self, other):
+        now = datetime.now()
+        def cond_eq(a, b):
+            return a is None or b is None or a == b
+        return (
+            isinstance(other, Dataset)
+            and (
+                # either both None or both datetime
+                not (bool(self.expires is None) ^ bool(other.expires is None))
+                # one none implies both none
+                and (self.expires is None or self.expires > now and other.expires > now)
+            )
+            and self.access_location == other.access_location
+            and self.category == other.category
+            and self.created_on == other.created_on
+            and self.data_domain == other.data_domain
+            and self.dataset_type == other.dataset_type
+            and self.is_read_only == other.is_read_only
+            and self.name == other.name
+            and cond_eq(self.derived_from, other.derived_from)
+            and cond_eq(self.last_updated, other.last_updated)
+            and cond_eq(self.uuid, other.uuid)
+        )
 
     @property
     def manager(self) -> Optional[DatasetManager]:
+        """
+        The ::class:`DatasetManager` for this instance.
+
+        Returns
+        -------
+        DatasetManager
+            The ::class:`DatasetManager` for this instance.
+        """
         return self._manager
 
-    @manager.setter
-    def manager(self, value: DatasetManager = None):
+    def set_manager(self, value: Union[DatasetManager, None]):
+        """
+        Sets the ::class:`DatasetManager` and updates the
+        ::attribute:`manager_uuid` property on this instances.
+
+        Parameters
+        ----------
+        value: DatasetManager | None
+            The new value for ::attribute:`manager`.
+        """
         self._manager = value if isinstance(value, DatasetManager) else None
 
         if value is not None:
```

Comment:
I think we need to do this to make sure `manager_uuid` also gets properly unset.

```suggestion
        if value is None:
            self.manager_uuid = None
        else:
```

Reply:
Thanks for pointing this out, @robertbartel. Just pushed up a fix. I slightly altered the possible behavior, but I think its for the best. LMK if you think otherwise.

Reply:
Hmmm, so what about the case when you are initializing from a deserialized json message that includes `manager_uuid`? This boils down to is the invariant that an instance's `manager` and `manager_uuid` fields are always either `None` or some valid value? Or can you have a `manager_uuid` field without a `manager` being set? I think the answer if for sure the latter.

Reply:
FWIW the previous behavior pre-pydantic was to leave the `manager_uuid` field if you set the manager with a `None` value (or any non-covariant `DatasetManager` type).

Reply:
> Hmmm, so what about the case when you are initializing from a deserialized json message that includes manager_uuid? 

Assuming we keep the `manager_uuid` attribute (I'll come back to this), we need to support this case.  The motive for including the `manager_uuid` field is that we can't serialize the manager, so instead we serialize a way to identify the right manager. 

> This boils down to is the invariant that an instance's manager and manager_uuid fields are always either None or some valid value?

Not strictly always.  Practically, but not strictly.  E.g., not immediately after deserialization.

> FWIW the previous behavior pre-pydantic was to leave the manager_uuid field if you set the manager with a None value (or any non-covariant DatasetManager type).

Right, I thought earlier in the discussion for the PR that this was an oversight.  The only reason to support setting the manager to `None`, beyond when we are first initializing the object, would be to unset the manager.  And we aren't properly unsetting the manager if the _Dataset_, when serialized, still reflects said manager's uuid.  For now though, leave this as you have it; I'm starting to think it more a consequence of a problem in the design than an oversight.

I don't want us to get further caught up here, so I'll open a separate issue, but I think we should examine whether it is really appropriate/necessary/useful to serialize the manager's uuid with a dataset.  Manager instance uuids end up being dynamically generated each time, and the service only permits one manager per _DatasetType_ (i.e., there's only ever one `OBJECT_STORE` type manager at a time, which gets an arbitrary uuid, so if we know a _Dataset_ is `OBJECT_STORE`, maybe we don't need to and shouldn't keep a manager uuid separately from a manager reference).